### PR TITLE
Clear OSF when setting time

### PR DESCRIPTION
**Pull request by danielspofford**
Why
---

- The OSF gets set to `1` in the following cases:
  - The first time power is applied.
  - Vcc and Vbat voltage cannot support oscillation.
  - The EOSC bit is turned off in battery-backed mode.
  - External influences on the crystal (i.e., noise, leakage, etc.).
- ...The OSF will remain set until explicitly written to. The only safe
  time to do this is when an external time is provided (NTP, etc.).

How
---

- `NervesTime.RTC.DS3231.set_time/2` now reads the status register,
  clears the OSF, and writes the status register only after successfully
  writing the date register.
- Add `NervesTime.RTC.DS3231.Status.encode/1`.

Side effects
------------

- Remove private
  `NervesTime.RTC.DS3231.{rtc_available?/2, supported?/1}`. These were
  only being used when `NervesTime` would try to initialize this
  library. The only check they were performing was whether the OSF was
  set, if it was, this library would fail to start and crash
  `NervesTime`. This is undesirable due to the above text with respect
  to the OSF - we need to be able to leverage `NervesTime` and likely
  NTP to obtain a valid time, write it to the RTC, and clear the OSF.

### CODE REVIEW

File: test/nerves_time/rtc/ds3231/status_test.exs [hunk 1]
```diff
@@ -2,9 +2,17 @@ defmodule NervesTime.RTC.DS3231.StatusTest do
...
+  test "decode/1 and encode/1" do
+    data = %{
+      alarm_1_flag: 1,
+      alarm_2_flag: 1,
+      busy: 1,
+      ena_32khz_out: 1,
+      osc_stop_flag: 1
+    }
+
+    bin = <<0x8F>>
```
Comment (danielspofford) [inline test/nerves_time/rtc/ds3231/status_test.exs]: Just pointing out that this changed from an `0xFF` above because 3 bits get correctly thrown away when decoding. In other words, previously `Status.decode(<<0xFF>>)` would yield `data` above. But so would all of these `["8F", "9F", "AF", "BF", "CF", "DF", "EF", "FF"]`. So now that we are also _encoding_ it does matter, because `<<0b1111_1111>>` encoded and then decoded would be.... `<<0x8F>>`.

TLDR: `Status.decode/1` is (and was) lossy, but only on bits with no meaning.

**Comment by fhunleth** [comment]
@danielspofford Could you rebase? I enabled CircleCI and fixed formatting issues, but it seems to have caused conflicts in your PR. The good news is that I think you'll have a cleaner PR when you rebase.

**Comment by fhunleth** [comment]
@jsimmonds2 If you're good with the changes, then so am I. Please feel free to press the "Rebase and merge" button when you're ready. I think that you should have permission. 

Also, it looks like a release could be made to hex. I'm certainly ok with you making it, but if you want me to do make the first one, just let me know when.

**Comment by jsimmonds2** [comment]
@fhunleth I would prefer that you make a **hex** release . . way faster. I've never looked at how release artefacts are created and submitted to hex.pm !

**Comment by fhunleth** [comment]
I can make it. 

@danielspofford Dialyzer flagged the following error (due to this PR predating CI running, it wasn't flagged):

```
lib/nerves_time/rtc/ds3231/status.ex:15:contract_supertype
Type specification is a supertype of the success typing.

Function:
NervesTime.RTC.DS3231.Status.encode/1

Type specification:
@spec encode(map()) :: {:ok, registers()}

Success typing:
@spec encode(%{
  :alarm_1_flag => integer(),
  :alarm_2_flag => integer(),
  :busy => integer(),
  :ena_32khz_out => integer(),
  :osc_stop_flag => integer(),
  _ => _
}) :: {:ok, <<_::8>>}
```

Could you fix?

**Comment by fhunleth** [comment]
Also, and I suppose I should have looked harder, the check for the RTC was removed so if the RTC doesn't physically exist, the `init` call will still succeed. I don't know this part, so I don't know the appropriate check, but it seems like it should at least try to read a register on the DS3231.

**Comment by danielspofford** [comment]
I will address the dialyzer issues in upcoming PRs if still there. WRT RTC check on init see https://github.com/nerves-time/nerves_time_rtc_ds3231/issues/6.
